src/dune_rules/coq/coq_lib.ml

```diff
@@ -41,37 +36,107 @@ module Id = struct
-    ; use_stdlib : bool
-          (* whether this theory uses the stdlib, eventually set to false for all libs *)
-    ; src_root : Path.Build.t
-    ; obj_root : Path.Build.t
-    ; theories : (Loc.t * t) list Resolve.t
-    ; libraries : (Loc.t * Lib.t) list Resolve.t
-    ; theories_closure : t list Resolve.t Lazy.t
-    ; package : Package.t option
-    }
-end
-
-let name l = l.id.name
-
-let implicit l = l.implicit
-
-let src_root l = l.src_root
-
-let obj_root l = l.obj_root
+type t =
+  { boot : t option Resolve.t
+  ; library : library
+  ; id : Id.t
+  }
+
+and library =
+  | Library of
+      { loc : Loc.t
+      ; src_root : Path.Build.t
+      ; obj_root : Path.Build.t
+      ; theories : (Loc.t option * t) list Resolve.t
+      ; libraries : (Loc.t * Lib.t) list Resolve.t
+      ; theories_closure : t list Resolve.t Lazy.t
+      ; package : Package.t option
+      }
+  | Legacy of
+      { implicit : bool (* Only useful for the stdlib *)
+      ; installed_root : Path.t
+      }
 
-let libraries l = l.libraries
+let rec to_dyn t =
+  Dyn.record
+    [ ("boot", Resolve.to_dyn (Dyn.option to_dyn) t.boot)
+    ; ("library", library_to_dyn t.library)
+    ; ("id", Id.to_dyn t.id)
+    ]
+
+and library_to_dyn = function
+  | Library t ->
+    Dyn.variant "Library"
+      [ Dyn.record
+          [ ("loc", Loc.to_dyn t.loc)
+          ; ("src_root", Path.Build.to_dyn t.src_root)
+          ; ("obj_root", Path.Build.to_dyn t.obj_root)
+          ; ( "theories"
+            , Resolve.to_dyn
+                (Dyn.list (Dyn.pair (Dyn.option Loc.to_dyn) to_dyn))
+                t.theories )
+          ; ( "libraries"
+            , Resolve.to_dyn
+                (Dyn.list (Dyn.pair Loc.to_dyn Lib.to_dyn))
+                t.libraries )
+          ; ("theories_closure", Dyn.opaque "<lazy>")
+          ; ("package", Dyn.option Package.to_dyn t.package)
+          ]
+      ]
+  | Legacy t ->
+    Dyn.variant "Legacy"
+      [ Dyn.record
+          [ ("implicit", Dyn.Bool t.implicit)
+          ; ("installed_root", Path.to_dyn t.installed_root)
+          ]
+      ]
 
-let package l = l.package
+let name t = t.id.name
+
+let loc t =
+  match t.library with
+  | Library l -> Some l.loc
+  | Legacy _ -> None
+
+let implicit t =
+  match t.library with
+  | Library _ -> false
+  | Legacy t -> t.implicit
+
+let src_root t =
+  match t.library with
+  | Library l -> Some l.src_root
+  | Legacy _ -> None
+
+let obj_root t =
+  match t.library with
+  | Library l -> Some l.obj_root
+  | Legacy _ -> None
+
+let installed_root t =
+  match t.library with
+  | Library _ -> None
+  | Legacy l -> Some l.installed_root
+
+let root t =
+  match t.library with
+  | Library l -> Path.build l.src_root
+  | Legacy l -> l.installed_root
+
+let theories t =
+  match t.library with
+  | Library l -> l.theories
+  | Legacy _ -> Resolve.return []
+
+let libraries t =
+  match t.library with
+  | Library l -> l.libraries
+  | Legacy _ -> Resolve.return []
```

Comment:
Actually IMHO it is better if we expose the two kind of libs in the interface, so rules can know what kind of library they are dealing with. Reasoning: this distinction really matters at rule generation time.